labs/b6.md

```diff
@@ -0,0 +1,144 @@
+
+Your mission, should you choose to accept it, is to write a systemd service that manages this web server. To do this, make a new unit file in `/etc/systemd/system/toy.service`. Refer to the slides for an example; DigitalOcean also has a [good guide](https://www.digitalocean.com/community/tutorials/understanding-systemd-units-and-unit-files) on how to write systemd units.
+
+Some fields worth considering for this service file
+- Description: give it a short description of what this unit is
+- Requires, After, Before: what units are necessary for our webserver to start, and do these units need to finish loading before the webserver starts (Hint: we probably want to ensure the network is online)?
+- ExecStart: which script should systemd run to start the webserver?
+ - WantedBy: at which stage of the booting process should this service be launched (Hint: `multi-user.target`)?
+ - User: The default is root. Is that a safe practice for a web servers?
+
+You are encouraged to experiment with other fields as suits your liking.
+
+Once you have finished creating `toy.service`, let's start the service and have the it start whenever our machine is booted.
+```
+# systemctl start toy.service
+# systemctl enable toy.service
+```
+
+### Debugging
+You can check if the unit file succeeded by running `systemctl status toy.service`. If you are having issues with the unit file or the web server, check the logs for this unit by running `journalctl -u toy.service`.  If you run into errors don't get demoralized (it is, after all, only a decal); as a sysadmin you'll have to become comfortable making sense of arcane error messages.
+
+### Crash the service!
+One of the great benefits of using systemd to manage your services is that you don't have to worry unnecessarily about bringing a process back up if it crashes. So let's crash the service! You can do this by either sending a POST request with the json payload `'{"crash":"true"}'` to http://yourvm.decal.xcf.sh:5000/crash (Hint: use `cURL`) or by kill the webserver manually by sending a signal -- both will cause the unit to crash. You can verify if you succeeded by running `systemctl status toy.service`, and the unit should either be in an `inactive` or `failed` state, depending on how you killed it. 
+
+Now add the following the `/etc/systemd/system/toy.service` under the `Service` directive:
+```
+Restart=always
+RestartSec=10
+```
+Now kill your webserver again in any way you please, and you should see that it come back online after 10 seconds!
+
+### Changing unit files without restart
+One last tidbit worth mentioning is that one is able to modify a systemd unit file while a service is running and apply the new changes without stopping or restarting the service. For example, change `RestartSec=2` and run `sudo systemctl daemon-reload`. If you kill the webserver one last time you will notice it takes only 2 seconds to come online now.
```

Comment:
You always need to run `systemctl daemon-reload`, even when the service is not running.